[RDY] Problem definition

gfokkema opened the pull request.

gfokkema committed:
Add problem definition, referincing previous sections.

Balletie commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,32 @@
 \label{sec:problem-definition}
 
+In the first section the scope of Spoofax has been explored.
+In the two subsequent sections REPL's and literate programming have been explained.
```
"REPLs"

Balletie commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,32 @@
+Besides exposing already existing Spoofax functionality in an interactive shell,
+our product should also expose additional features geared towards interactive
+exploration of programs in the generated languages.
+Examples include displaying and editing the current programming environment,
```
what is "displaying and editing the current programming environment"?

Balletie commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,32 @@
+for literate programming, allowing for rapid prototyping and documentation at
+the same time~\cite{org-mode}. This would allow developers of new DSLs to
+document and explain their language in an interactive manner that directly
+allows experimentation.
```
nice

Balletie commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,32 @@
 
+In the first section the scope of Spoofax has been explored.
+In the two subsequent sections REPL's and literate programming have been explained.
+Given this information the problem definition can be adequately composed.
```
I don't think it's necessary to repeat what the previous sections were about. I think the purpose this intro is to just say that the background information has been discussed, and that therefore we can now discuss the problem itself. So I think you should just say something like that. For example:

> The previous (three) sections gave the necessary background knowledge of the problem domain. Now that the background knowledge has been discussed, this section can give the problem definition of the client.

Balletie commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,32 @@
+Given this information the problem definition can be adequately composed.
+
+Spoofax offers a wide variety of tools to develop definitions of domain specific
+languages. Given these definitions, Spoofax generates several DSL specific tools.
```
"Develop DSLs" should be enough.

I also think that it doesn't have to be explained again that Spoofax generates stuff from language definitions

Maybe:

> Section 3 has shown that Spoofax offers a wide variety of tools to develop DSLs and accompanying IDE support.

Balletie commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,32 @@
+Precisely because Spoofax is concerned with developing DSLs, rapid prototyping
+of syntax and grammar is a convenient addition to the product.
+In section 2 REPLs were explored, and it was explained that these provide exactly
+this rapid prototyping ability for other languages like LISP and Python.
```
Use keywords such as "Therefore", so that there is a flow in the story about the origin of the problem.

Try not to repeat too much: "In section 2 REPLs were explored" has been mentioned a couple of times. It is enough to say that "section 2 showed that REPLs provide rapid prototyping ability", or something similar.

Maybe something like:

> Section 2 showed that REPLs provide precisely this rapid prototyping ability for other languages like LISP and Python. Therefore, the client has expressed their interest in REPLs that are generated for languages developed with Spoofax. Such a REPL would aid both the developer of the language, and its end-user.

gfokkema committed:
Fix Skip's comments.

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,33 @@
 
+The previous sections provided the necessary background knowledge of the problem
+domain. Now that the background knowledge has been discussed,
+this section will give the problem definition of the client.
```
`give` -> `discuss`?

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,33 @@
+Spoofax offers a wide variety of tools to develop DSLs and accompanying IDE support.
+Precisely because Spoofax is concerned with developing DSLs, rapid prototyping
+of syntax and grammar is a convenient addition to the product.
+Section 2 showed that REPLs provide precisely this rapid prototyping ability
```
Reference?

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,33 @@
+Precisely because Spoofax is concerned with developing DSLs, rapid prototyping
+of syntax and grammar is a convenient addition to the product.
+Section 2 showed that REPLs provide precisely this rapid prototyping ability
+for other languages like LISP and Python. Therefore, the client has expressed
```
`Lisp`. LISP is the historically used name, it's 2016 now :P In all seriousness though, section 2 uses Lisp so we should be consistent.

Hjdskes replied:
I don't think you need to name example languages though.

Balletie replied:
I think it's nice. You can read it like: "It works for other languages, so it should work for languages made in spoofax too"

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,33 @@
+
+Since a lot of language services are already provided by Spoofax,
+the REPL should try to hook in to existing services as much as possible.
+Besides exposing already existing Spoofax functionality in an interactive shell,
```
What do you mean with `existing Spoofax functionality`?

Balletie replied:
I think he means the existing services mentioned in the previous sentence

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,33 @@
+Since a lot of language services are already provided by Spoofax,
+the REPL should try to hook in to existing services as much as possible.
+Besides exposing already existing Spoofax functionality in an interactive shell,
+our product should also expose additional features geared towards interactive
```
Don't refer to ourselves in a formal document. `our` -> `the`.

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,33 @@
+
+When a REPL is realized, the product could be further extended to add support
+for literate programming, allowing for rapid prototyping and documentation at
+the same time~\cite{org-mode}. This would allow developers of new DSLs to
```
Org-mode isn't really used for programming though, right? In that sense, it might not be the best example to use.

Balletie replied:
This is not true. There are people whose entire emacs configuration is an org-mode document, for example. Moreover, it is something which you don't necessarily "see", as the source code is /tangled/ out, and the documentation is /weaved/ out.

Balletie committed:
Process feedback Jente. Also "in to" -> "into"

Word-diff, because line noise due to formatting: http://i.imgur.com/kbK1PeD.png

Balletie commented:
I processed jente's feedback. Word-diff: http://i.imgur.com/kbK1PeD.png

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,36 @@
 
+The previous sections provided the necessary background knowledge of
+the problem domain. Now that the background knowledge has been
+discussed, this section will discuss the problem definition of the
```
Two times `discuss`

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,36 @@
+accompanying IDE support. Precisely because Spoofax is concerned with
+developing DSLs, rapid prototyping of syntax and grammar is a
+convenient addition to the product. \Cref{sec:read-eval-print} showed
+that REPLs provide precisely this rapid prototyping ability for other
```
Again `precisely`, already used in previous sentence.

Hjdskes commented on research/04-problem-def.tex:
```diff
@@ -1,6 +1,36 @@
+shell, the product should also expose additional features geared
+towards interactive exploration of programs in the generated
+languages.  Examples include displaying and editing the contents of
+current program context, saving and loading program context and
```
Forgot `the`

gfokkema replied:
*left out `the` :p

gfokkema committed:
Fix Jente's last comments.
